scripts/build/builders/nuttx.py

```diff
@@ -0,0 +1,137 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import os
+from enum import Enum, auto
+
+from .gn import Builder
+
+
+class NuttXApp(Enum):
+    LIGHT = auto()
+
+    def ExampleName(self):
+        if self == NuttXApp.LIGHT:
+            return 'lighting-app'
+        else:
+            raise Exception('Unknown app type: %r' % self)
+
+    def AppNamePrefix(self, chip_name):
+        if self == NuttXApp.LIGHT:
+            return ('chip-%s-lighting-example' % chip_name)
+        else:
+            raise Exception('Unknown app type: %r' % self)
+
+
+class NuttXBoard(Enum):
+    SIM = auto()
+
+
+def NuttXTarget(board, app):
+    if board == NuttXBoard.SIM:
+        if app == NuttXApp.LIGHT:
+            return 'sim:matter'
+
+
+class NuttXBuilder(Builder):
+
+    def __init__(self,
+                 root,
+                 runner,
+                 app: NuttXApp = NuttXApp.LIGHT,
+                 board: NuttXBoard = NuttXBoard.SIM,
+                 enable_rpcs: bool = False,
+                 module_type: str = 'sim',
+                 baudrate=2000000,
+                 enable_shell: bool = False,
+                 enable_cdc: bool = False,
+                 enable_resetCnt: bool = False,
+                 enable_rotating_device_id: bool = False,
+                 enable_ethernet: bool = False,
+                 enable_wifi: bool = False,
+                 enable_thread: bool = False,
+                 enable_frame_ptr: bool = False,
+                 enable_heap_monitoring: bool = False,
+                 use_matter_openthread: bool = False,
+                 ):
+
+        nuttx_chip = 'nuttx'
+
+        super(NuttXBuilder, self).__init__(
+            root=os.path.join(root, 'examples',
+                              app.ExampleName(), nuttx_chip),
+            runner=runner
+        )
+
+        self.argsOpt = []
+        self.chip_name = nuttx_chip
+        self.enable_frame_ptr = enable_frame_ptr
+        self.app = app
+        self.board = board
+        # self.nuttx_dir = os.path.join(self.output_dir, 'nuttx')
+        # self.nuttx_apps_dir = os.path.join(self.output_dir, 'apps')
+
+    def print_enviroment_error(self):
+        logging.fatal('*' * 80)
+        logging.fatal('*' * 80)
+
+    def generate(self):
+        logging.info('generate NuttX %s at %s',
+                     NuttXTarget(self.board, self.app), self.output_dir)
+
+        self._Execute(['mkdir', '-p', self.output_dir],
+                      title='Generating ' + self.identifier)
+
+        nuttx_dir = os.path.join(os.sep, 'opt', 'nuttx', 'nuttx')
+        self._Execute(['sed', '-i', 's/set(CMAKE_CXX_STANDARD 14)//g', os.path.join(nuttx_dir, 'CMakeLists.txt')])
+        config_path = os.path.join(nuttx_dir, 'boards', 'sim', 'sim', 'sim', 'configs', 'matter', 'defconfig')
+        self._Execute(['sed', '-i', 's/CONFIG_SIM_M32=y//g', config_path])
```

Comment:
Could you explain these SEDs ? These seem to dirty/change some global locations. If these are needed, they should probably be some environment/SDK setup scripts instead rather than part of the build system. Generaly I expect builds to only ever touch files inside the output directory and never change the SDK or source tree.

Reply:
@andy31415 . I encountered two minor issues when trying to compile Matter in my local Docker environment. These issues can be easily avoided using sed, so I added the relevant sed commands to the script. The corresponding issues have already been fixed and merged into the master branch of NuttX. If there are new features that require an update to the NuttX SDK in the future, the relevant modifications will also be included. I'm not sure if this is allowed.

Reply:
These seds are not allowed as part of the build script. The build scripts are not allowed to dirty neither the source code nor SDKs.

you are only allowed to touch files inside the build output directory (e.g. esp32 builds needed sdkconfig files ... so what they do is copy those files into the build directory and change them there.

Reply:
If you need to alter the SDKs, you should do that as part of the docker file (likely using some form of separate script so that users can apply equivalent changes to their local non-docker environments)

Reply:
Done.